### ISSUE TITLE
Vine: Memory Leak in FunctionCall Tasks

### ISSUE THREAD

**Issue by JinZhou5042** [cooperative-computing-lab/cctools#3572]
Here is my simple example using serverless function calls to perform random matrix multiplications:
```python
import ndcctools.taskvine as vine
from tqdm import tqdm

def random_matrix_multiplication(size=5):
    import numpy as np
    matrix1 = np.random.rand(size, size)
    matrix2 = np.random.rand(size, size)
    result_matrix = np.dot(matrix1, matrix2)

    return result_matrix


def main(): 
    q = vine.Manager(port=9123)
    q.set_name("test_manager")
    print(f"TaskVine manager listening on port {q.port}")

    print("Creating library from packages and functions...")
    import_modules = []
    libtask = q.create_library_from_functions('test-library', random_matrix_multiplication, import_modules=import_modules)

    q.install_library(libtask)

    print("Submitting function call tasks...")

    tasks = 50000

    for _ in range(0, tasks):
        s_task = vine.FunctionCall('test-library', 'random_matrix_multiplication', 1000)
        q.submit(s_task)

    print("Waiting for results...")

    pbar = tqdm(total=tasks)
    while not q.empty():
        t = q.wait(0)
        if t:
            pbar.update(1)

    pbar.close()


if __name__ == '__main__':
    main()
```

Here is my `vine_factory` configuration:
```json
{
    "manager-name": "test_manager",
    "max-workers": 512,
    "min-workers": 512,
    "workers-per-cycle": 512,
    "cores": 1,
    "memory": 1024,
    "timeout": 36000
}
```
It works well with small amount of tasks, but failed somehow as I increased the number of tasks:

![Screenshot from 2023-11-16 10-30-31](https://github.com/cooperative-computing-lab/cctools/assets/142265839/7d08ff58-1c5a-4f1c-a0e6-69f0db805a68)
![Screenshot from 2023-11-16 10-30-54](https://github.com/cooperative-computing-lab/cctools/assets/142265839/7dff8847-ffc0-4000-affa-9f58903c7f39)

I am trying different configurations to look into this problem.

**Comment by dthain** [comment]
Take a good look at the assertion failure, it tells you exactly where things are going wrong:
https://github.com/cooperative-computing-lab/cctools/blob/05d31a9cb5ed77af37d0c6e8b0136fd81b0e94c4/taskvine/src/manager/vine_manager.c#L2836

Suggest that you start by running the manager with `valgrind` to make sure there isn't some other memory problem going on.

Otherwise, there is some subtle inconsistency between `t->worker` and `w->current_tasks`

**Comment by JinZhou5042** [comment]
There might be some memory leak problem, each time a task is completed, some memory was leaked, and the program stops when significant memory is lost.

<img width="1298" alt="image" src="https://github.com/cooperative-computing-lab/cctools/assets/142265839/605aa46e-09a9-407a-93d1-7850757023bc">

**Comment by dthain** [comment]
Indeed!  Is it the manager or the worker running out of memory?

**Comment by JinZhou5042** [comment]
I'm still figuring it out. I'll submit only several tasks and workers and see who is dirtying our memory.

**Comment by JinZhou5042** [comment]
I am running this script and using `htop` to monitoring the real-time resources consuming, this is tasks running status:

![Screenshot from 2023-11-29 17-14-25](https://github.com/cooperative-computing-lab/cctools/assets/142265839/1908eac3-d29e-4952-ab34-63e37d41d268)

And the consumed resources:

![Screenshot from 2023-11-29 17-14-39](https://github.com/cooperative-computing-lab/cctools/assets/142265839/8cd2b3eb-51ca-4b74-8ed7-e16c28e0e4a7)

The manager has consumed around 400G!

I did the same thing on my local desktop, just using pure Python for loop to run the function, let's see the running status and consumed resources:

![Untitled](https://github.com/cooperative-computing-lab/cctools/assets/142265839/6561c615-5bca-47ea-9635-abf1af1706a1)

![Untitled(1)](https://github.com/cooperative-computing-lab/cctools/assets/142265839/c7341fba-8d3d-4f95-8fe9-412505a55a79)

More tasks are completed, but only 230M memory was used.

**Comment by tphung3** [comment]
It looks like we have a memory leak when tasks are submitted as function calls. I get this by running a script submitting 10K tasks, each simply returning a list with 100K elements. For the same amount of tasks, the manager process uses 70KBs of memory at all times if tasks are submitted as PythonTask. However if tasks are submitted as FunctionCalls, then the manager process uses up to 2GBs of memory and its memory usage grows gradually.

My hypothesis is that there's a bug with how we track and delete outputs from function calls. One possible way is that the outputs are referenced elsewhere in the code and thus not garbage collected.

@JinZhou5042 Is it similar to your observation above?

**Comment by JinZhou5042** [comment]
@dthain Exactly!

**Comment by JinZhou5042** [comment]
@tphung3 Exactly!

**Comment by dthain** [comment]
Ok, you are making progress on tracking this down.  Next thing to try:
Write the same benchmark in C and Python and see if both exhibit the problem.
That will tell you whether it is in the core vine_manager, or in the Python bindings along.

**Comment by tphung3** [comment]
This issue might also come from the difference in how PythonTask and FunctionCall are implemented. Result from a PythonTask is returned to the manager as a file, and it's only loaded into the memory once the user calls `t.output`. FunctionCall instead returns result via a TaskVine buffer, so it's loaded into the memory no matter what. In the experiment I described above, I never called `t.output` and instead just loop through to get all tasks' results to come back, resulting in the observation above.

Just to add on to @dthain 's suggestion, I think we need to think more carefully whether this is a design issue or not, and then proceed with a solution.

**Comment by dthain** [comment]
In our discussion yesterday, I observed that `FunctionCallTask` invokes `m.declare_buffer` twice (one for the input object, and once for the output, but nobody ever removes them from the manager.

That has to be at least part of the problem.

Since the class was responsible for generating them (and nobody will ever access those directly) I think the class should also be responsible for deleting them.

**Comment by dthain** [comment]
@JinZhou5042 and @BarrySlyDelgado discuss this and come up with a solution.

As Barry points out, it is complicated slightly by the idea of "distant futures" in which a later function call may want to consume the output from a prior function call.

**Comment by BarrySlyDelgado** [comment]
Once we retrieve the output, the files may not be needed anymore. So the opportunity to delete these files may be when t.output is called.

**Comment by dthain** [comment]
Let me propose the general rule that whatever layer created a file should be responsible for deleting the file.

In this case the FunctionCall object itself is creating a file object (that the user doesn't know about!) and so the object itself should be responsible for deleting it.

**Comment by dthain** [comment]
I don't think you should be calling `_free` explicitly. 
That's just used as the final cleanup prior to garbage collection, right?
Just remove the file objects from the manager, and that should be enough.

**Comment by JinZhou5042** [comment]
That was a little bit tricky... the `_input_buffer` was referenced twice, the `m->file_table` and the `t->input_mounts`.

No matter how many times I call `self._manager.remove_file(self._input_buffer)`, it only removes the reference on `m->file->table`.

I didn't find a function in `vine_task.c` that explicitly removes its  mounted nodes, and therefore I called `self._free()`.

Should we add a new function in `vine_task.c` to remove `t->input_mounts` or `t->output_mounts`?

**Comment by dthain** [comment]
Please just make the FunctionCall call `vine_remove_file` once for each `vine_declare_file`.  That's all you have to do.

When the FunctionCall is garbage collected, it will delete its underlying task and files in the proper way.  Each object is responsible for deleting what it has created, no more and no less.

The rest of the cleanup works as designed:
- FunctionCall is a superclass of Task, which called `vine_task_create` in its constructor, and calls `vine_task_delete` in its destructor.
- `vine_task_delete` deletes all of its mounts with `vine_mount_delete`
- `vine_mount_delete` deletes its own file with `vine_file_delete`

Likewise:
- `vine_declare_file` creates a file object, adds it to the `file_table` and adds a reference.
- `vine_remove_file` removes a file object from the table, and deletes a reference.

Every create/clone has a mirror image delete.  You are only responsible for deleting what you create, no more and no less.

**Comment by dthain** [comment]
Good, we are making progress.
Come on over and lets talk a bit.

**Comment by JinZhou5042** [comment]
1. I invoked `gc.enable()` and `gc.collect()` in various places in my example, didn't change its current behaviors.
2. I ran 1000 tasks, with no object was garbage collected in this course, in total 67.33GB memory was consumed:
![Screenshot from 2023-12-13 16-21-56](https://github.com/cooperative-computing-lab/cctools/assets/142265839/2f06ac11-ebe4-4f6c-b617-77cf43886920)

![Screenshot from 2023-12-13 16-22-35](https://github.com/cooperative-computing-lab/cctools/assets/142265839/bd36dce5-0430-4158-bf83-5446143fb488)

**Comment by JinZhou5042** [comment]
The condition of 10000 tasks:

![Screenshot from 2023-12-13 18-06-07](https://github.com/cooperative-computing-lab/cctools/assets/142265839/0fd878a6-ac14-413d-a967-ece54caed20e)

**Comment by JinZhou5042** [comment]
@dthain @btovar  Great news! I found it is the `self._finalizer = weakref.finalize(self, self._free)` in `class Task` that keeps a "permanent" reference to  the task object, this reference was removed at the very end of our program. I need to familiarize my self more about how `weakref.finalize` works.

After commenting this line of code, and explicitly adding a common destructor function in `Task`:
``` Python
def __del__(self):
    self._free()
```
The reference count of `t` goes back to `1`, which is what we want, the destructor can be called properly without additional reference, and our destructor can be invoked accordingly:

![Screenshot from 2023-12-14 13-44-17](https://github.com/cooperative-computing-lab/cctools/assets/142265839/3c13b1f7-d635-4c62-9729-e52acc092837)
